ambry-store/src/main/java/com.github.ambry.store/CompactionDetails.java

```diff
@@ -0,0 +1,165 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * Contains all the details required for a compaction cycle.
+ */
+class CompactionDetails {
+  private static final byte[] ZERO_LENGTH_ARRAY = new byte[0];
+
+  private static final short VERSION_0 = 0;
+  private static final int VERSION_SIZE = 2;
+  private static final int REFERENCE_TIME_SIZE = 8;
+  private static final int SEGMENT_COUNT_SIZE = 4;
+  private static final int SEGMENT_NAME_LENGTH_SIZE = 4;
+  private static final int SWAP_SPACE_COUNT_SIZE = 4;
+
+  private final long referenceTime;
+  private final List<String> logSegmentsUnderCompaction;
+  private final String extraSegmentName;
+  private final int swapSpaceCount;
+
+  /**
+   * Construct a representation of all the details required for a compaction cycle.
+   * @param referenceTime the epoch time to use to get the same results from the {@link BlobStore} as when these details
+   *                      were created.
+   * @param logSegmentsUnderCompaction the names of the {@link LogSegment} under compaction.
+   * @param extraSegmentName the name of the segment which will provide the "extra" data. Can be {@code null} or empty
+   *                         if no extra data is required.
+   * @param swapSpaceCount the number of swap spaces that will be required for compaction.
+   */
+  CompactionDetails(long referenceTime, List<String> logSegmentsUnderCompaction, String extraSegmentName,
+      int swapSpaceCount) {
+    this.referenceTime = referenceTime;
+    this.logSegmentsUnderCompaction = logSegmentsUnderCompaction;
+    this.extraSegmentName = extraSegmentName;
+    this.swapSpaceCount = swapSpaceCount;
+  }
+
+  /**
+   * @param stream the serialized version of the {@link CompactionDetails} that has to be loaded.
+   * @return a representation of all the details required for compaction from the given {@code stream}.
+   * @throws IllegalArgumentException if the version is not recognized
+   * @throws IOException if there is an I/O error reading from the stream
+   */
+  static CompactionDetails fromBytes(DataInputStream stream) throws IOException {
+    CompactionDetails details;
+    short version = stream.readShort();
+    switch (version) {
+      case VERSION_0:
+        long referenceTime = stream.readLong();
+        int segmentCount = stream.readInt();
+        List<String> logSegmentsUnderCompaction = new ArrayList<>();
+        for (int i = 0; i < segmentCount; i++) {
+          logSegmentsUnderCompaction.add(Utils.readIntString(stream));
+        }
+        String extraLogSegmentName = Utils.readIntString(stream);
+        if (extraLogSegmentName != null && extraLogSegmentName.isEmpty()) {
+          extraLogSegmentName = null;
```

Comment:
Good point - I removed the javadoc that says it can be empty. It is easier to reason and handle in code if it is one of the two (`null` or empty) so I have gone with expecting it to be `null`if no extra segment is required.